File: PHASE::TEST/keygen_oracle.py
# -*- coding: utf-8 -*-
# @category Validation
# @runtime Jython
#
# Live emulation harness for synthetic_keygen.py validation.
# Loads synthetic_tc1766.bin in Ghidra. Uses SYNTHETIC polynomial table.
# Drives FUN_8001b57e (the LFSR seed/key primitive at 0x8001b57e) for several
# (subfunc, STM_TIM0) combinations. For each trial it captures:
#   - the firmware-generated seed (bytes the dispatcher would put at tx_buf+0x1C)
#   - the firmware-computed expected_key (stored to *(a0 - 0x581c))
# Output is JSON, written to /tmp/keygen_oracle_trials.json, so an external
# step can feed each seed into synthetic_keygen.py and compare with expected_key.
#
# Nothing is mocked, simulated by hand, or symbolic. The firmware bytes are
# executed; the function's own outputs are what we record.

from ghidra.app.emulator import EmulatorHelper
from java.math import BigInteger
import json

# -- target & calling convention --
FUN_LFSR     = 0x8001b57e
SENTINEL     = 0xDEADBEEE         # even -> valid TriCore PC, used as halt PC

# a1 is the small-data base used to index the polynomial table:
#   poly_base = a1 - 0x7eca = 0x80025B6E   =>   a1 = 0x8002DA38
A1_VAL       = 0x8002DA38

# a0 is the LDRAM small-data base. We pick a value such that every a0+disp
# referenced by FUN_8001b57e lands in mapped LDRAM (0xD0000000..0xD000DFFF).
# Offsets used by the function: -0x55c0, -0x55e0, -0x558c, -0x56fc, -0x581c.
# Setting a0 = 0xD000C000 puts every reference into 0xD00067E4..0xD0006A74.
A0_VAL          = 0xD000C000
EXPECTED_KEY_AD = A0_VAL - 0x581c   # 0xD00067E4
SAVED_STATE_AD  = A0_VAL - 0x56fc   # 0xD0006904
UNLOCK_BASE     = A0_VAL - 0x55e0   # 0xD0006A20  (level byte stored at +0x54)
STATE_FLAG_AD   = A0_VAL - 0x55c0   # 0xD0006A40

SEED_BUF        = 0xD0008000        # passed as param_1 (a4): seed written here
STACK_TOP       = 0xD000B000
STM_TIM0        = 0xF0000210

space = currentProgram.getAddressFactory().getDefaultAddressSpace()

def A(off):
    return space.getAddress(off & 0xFFFFFFFF)

def bi(v):
    return BigInteger.valueOf(v & 0xFFFFFFFF)

def write_u32_le(emu, off, value):
    b = bytearray(4)
    b[0] = (value      ) & 0xFF
    b[1] = (value >>  8) & 0xFF
    b[2] = (value >> 16) & 0xFF
    b[3] = (value >> 24) & 0xFF
    emu.writeMemory(A(off), bytes(b))

def read_u32_le(emu, off):
    raw = emu.readMemory(A(off), 4)
    return (raw[0] & 0xFF) | ((raw[1] & 0xFF) << 8) | ((raw[2] & 0xFF) << 16) | ((raw[3] & 0xFF) << 24)

def read_bytes(emu, off, n):
    raw = emu.readMemory(A(off), n)
    return bytearray([(raw[i] & 0xFF) for i in range(n)])

def to_uint(v):
    # Jython auto-unwraps BigInteger to a Python long; handle both cases.
    if hasattr(v, "longValue"):
        return v.longValue() & 0xFFFFFFFF
    return int(v) & 0xFFFFFFFF

def run_trial(emu, subfunc, timer_val, saved_state=0, loop_modifier=0):
    # Reset relevant globals and registers
    emu.writeRegister("a0",  bi(A0_VAL))
    emu.writeRegister("a1",  bi(A1_VAL))
    emu.writeRegister("a4",  bi(SEED_BUF))
    emu.writeRegister("a10", bi(STACK_TOP))
    emu.writeRegister("a11", bi(SENTINEL))
    emu.writeRegister("d4",  bi(subfunc))
    emu.writeRegister("d5",  bi(loop_modifier))
    emu.writeRegister("pc",  bi(FUN_LFSR))

    write_u32_le(emu, SAVED_STATE_AD,  saved_state)
    write_u32_le(emu, EXPECTED_KEY_AD, 0)
    write_u32_le(emu, STATE_FLAG_AD,   0)
    write_u32_le(emu, SEED_BUF,        0)
    write_u32_le(emu, STM_TIM0,        timer_val)

    err = ""
    steps = 0
    MAX_STEPS = 200000
    while steps < MAX_STEPS:
        pc = to_uint(emu.readRegister("pc"))
        if pc == (SENTINEL & 0xFFFFFFFF):
            break
        if not emu.step(monitor):
            err = "step fault @ 0x%08x: %s" % (pc, emu.getLastError() or "?")
            break
        steps += 1
    else:
        err = "max steps exhausted"

    ret_code = to_uint(emu.readRegister("d2")) & 0xFF
    seed_b   = read_bytes(emu, SEED_BUF, 4)
    seed_be  = (seed_b[0] << 24) | (seed_b[1] << 16) | (seed_b[2] << 8) | seed_b[3]
    exp_key  = read_u32_le(emu, EXPECTED_KEY_AD)
    state_fl = read_u32_le(emu, STATE_FLAG_AD)

    return {
        "subfunc":      "0x%02X" % subfunc,
        "subfunc_int":  subfunc,
        "loop_modifier": loop_modifier,
        "timer":        "0x%08X" % timer_val,
        "saved_state":  "0x%08X" % saved_state,
        "ret_code":     "0x%02X" % ret_code,
        "seed_hex":     "%02X%02X%02X%02X" % (seed_b[0], seed_b[1], seed_b[2], seed_b[3]),
        "seed_be_u32":  "0x%08X" % seed_be,
        "expected_key": "0x%08X" % exp_key,
        "state_flag":   "0x%08X" % state_fl,
        "steps":        steps,
        "error":        err,
    }

# ---- trials ----
emu = EmulatorHelper(currentProgram)
trials = []
try:
    subfuncs = [0x01, 0x03, 0x05, 0x07, 0x09, 0x0B]
    # By varying saved_state we control the seed directly:
    #   seed = saved_state XOR poly_table[timer & 0x3F]
    # Timer fixed at 0 (=> XOR with poly[0]=0xA218B900 (synthetic) — predictable & valid).
    # The set below covers edge cases (all-zero, all-one, single-bit, mixed)
    # plus diverse arbitrary patterns.
    timer_val = 0
    seeds_via_saved_state = [
        0x00000000, 0x00000001, 0xFFFFFFFF, 0x80000000,
        0x12345678, 0xDEADBEEF, 0xCAFEBABE, 0xA5A5A5A5,
        0x5A5A5A5A, 0xAAAAAAAA, 0x55555555, 0x01010101,
        0x10101010, 0x7FFFFFFF, 0x00000002, 0x00000003,
        0xA218B900, 0x8F04FB9D, 0xF0529CFB, 0xAF83B3F7,
        0x98C02CAF, 0xE0E0B04A, 0xFF00FF00, 0x00FF00FF,
        0x0F0F0F0F, 0xF0F0F0F0, 0xDEADC0DE, 0xFACEFEED,
        0xBADC0FFE, 0x8BADF00D,
    ]
    POLY0 = 0xA218B900  # synthetic poly_table[0], XORed with saved_state when timer=0
    for sf in subfuncs:
        for s in seeds_via_saved_state:
            # We want the firmware to produce 'seed = s'.
            # seed = saved_state XOR poly[0], so saved_state = s XOR poly[0].
            ss = (s ^ POLY0) & 0xFFFFFFFF
            trials.append(run_trial(emu, sf, timer_val, saved_state=ss))
finally:
    emu.dispose()

out_path = "/tmp/keygen_oracle_trials.json"
f = open(out_path, "w")
f.write(json.dumps(trials, indent=2))
f.close()

print "wrote %d trials -> %s" % (len(trials), out_path)
for t in trials:
    print "[%s tim=%s] ret=%s seed=%s exp_key=%s err=%s" % (
        t["subfunc"], t["timer"], t["ret_code"], t["seed_hex"],
        t["expected_key"], t["error"] or "-")

File: firmware_src/ghidra_scripts/KeygenOracle.java
//@category Validation
//@runtime Java
//
// Java port of PHASE::TEST/keygen_oracle.py.
// Drives FUN_8001B57E in Ghidra's emulator with the rebuilt-binary bytes
// (bit-reversed Fibonacci LFSR formulation) over a sweep of
// (subfunc, saved_state) pairs and dumps (seed, expected_key) tuples to
// /tmp/keygen_oracle_trials.json so run_validation.py can cross-check the
// new synthetic_keygen.py.

import ghidra.app.script.GhidraScript;
import ghidra.app.emulator.EmulatorHelper;
import ghidra.program.model.address.Address;
import ghidra.program.model.address.AddressSpace;
import java.io.FileWriter;
import java.io.PrintWriter;
import java.math.BigInteger;
import java.util.ArrayList;
import java.util.List;

public class KeygenOracle extends GhidraScript {

    // Calling convention / memory layout (mirrors keygen_oracle.py)
    static final long FUN_LFSR        = 0x8001B57EL;
    static final long SENTINEL        = 0xDEADBEEEL;       // return-PC halt
    static final long A1_VAL          = 0x8002DA38L;       // poly-table base anchor
    static final long A0_VAL          = 0xD000C000L;       // small-data base
    static final long EXPECTED_KEY_AD = A0_VAL - 0x581cL;  // g_expected_key
    static final long SAVED_STATE_AD  = A0_VAL - 0x56fcL;  // g_rolling_seed
    static final long STATE_FLAG_AD   = A0_VAL - 0x55c0L;  // g_sec_state_flags
    static final long SEED_BUF        = 0xD0008000L;
    static final long STACK_TOP       = 0xD000B000L;
    static final long STM_TIM0        = 0xF0000210L;

    private AddressSpace defaultSpace;

    private Address A(long off) {
        return defaultSpace.getAddress(off & 0xFFFFFFFFL);
    }

    private void writeU32LE(EmulatorHelper emu, long off, long value) throws Exception {
        byte[] b = new byte[4];
        b[0] = (byte) (value & 0xFF);
        b[1] = (byte) ((value >> 8) & 0xFF);
        b[2] = (byte) ((value >> 16) & 0xFF);
        b[3] = (byte) ((value >> 24) & 0xFF);
        emu.writeMemory(A(off), b);
    }

    private long readU32LE(EmulatorHelper emu, long off) throws Exception {
        byte[] raw = emu.readMemory(A(off), 4);
        return ((raw[0] & 0xFFL))
             | ((raw[1] & 0xFFL) << 8)
             | ((raw[2] & 0xFFL) << 16)
             | ((raw[3] & 0xFFL) << 24);
    }

    private byte[] readBytes(EmulatorHelper emu, long off, int n) throws Exception {
        byte[] raw = emu.readMemory(A(off), n);
        byte[] out = new byte[n];
        for (int i = 0; i < n; i++) out[i] = (byte) (raw[i] & 0xFF);
        return out;
    }

    private long toUint(BigInteger v) {
        return v.longValue() & 0xFFFFFFFFL;
    }

    static class Trial {
        int    subfunc;
        long   timer;
        long   savedState;
        long   retCode;
        byte[] seedBytes;
        long   expectedKey;
        long   stateFlag;
        long   steps;
        String error;
    }

    private Trial runTrial(EmulatorHelper emu, int subfunc, long timerVal,
                           long savedState, long loopModifier) throws Exception {
        Trial t = new Trial();
        t.subfunc = subfunc;
        t.timer = timerVal;
        t.savedState = savedState;

        emu.writeRegister("a0",  BigInteger.valueOf(A0_VAL));
        emu.writeRegister("a1",  BigInteger.valueOf(A1_VAL));
        emu.writeRegister("a4",  BigInteger.valueOf(SEED_BUF));
        emu.writeRegister("a10", BigInteger.valueOf(STACK_TOP));
        emu.writeRegister("a11", BigInteger.valueOf(SENTINEL));
        emu.writeRegister("d4",  BigInteger.valueOf(subfunc));
        emu.writeRegister("d5",  BigInteger.valueOf(loopModifier));
        emu.writeRegister("pc",  BigInteger.valueOf(FUN_LFSR));

        writeU32LE(emu, SAVED_STATE_AD,  savedState);
        writeU32LE(emu, EXPECTED_KEY_AD, 0);
        writeU32LE(emu, STATE_FLAG_AD,   0);
        writeU32LE(emu, SEED_BUF,        0);
        writeU32LE(emu, STM_TIM0,        timerVal);

        long steps = 0;
        final long MAX_STEPS = 200000;
        while (steps < MAX_STEPS) {
            long pc = toUint(emu.readRegister("pc"));
            if (pc == (SENTINEL & 0xFFFFFFFFL)) break;
            if (!emu.step(monitor)) {
                t.error = String.format("step fault @ 0x%08x: %s",
                                        pc, emu.getLastError());
                break;
            }
            steps++;
        }
        if (steps >= MAX_STEPS) t.error = "max steps exhausted";

        t.retCode     = toUint(emu.readRegister("d2")) & 0xFF;
        t.seedBytes   = readBytes(emu, SEED_BUF, 4);
        t.expectedKey = readU32LE(emu, EXPECTED_KEY_AD);
        t.stateFlag   = readU32LE(emu, STATE_FLAG_AD);
        t.steps       = steps;
        return t;
    }

    private String hexBytes(byte[] b) {
        StringBuilder sb = new StringBuilder();
        for (byte x : b) sb.append(String.format("%02X", x & 0xFF));
        return sb.toString();
    }

    private void writeJson(List<Trial> trials, String path) throws Exception {
        try (PrintWriter pw = new PrintWriter(new FileWriter(path))) {
            pw.println("[");
            for (int i = 0; i < trials.size(); i++) {
                Trial t = trials.get(i);
                pw.println("  {");
                pw.printf("    \"subfunc\": \"0x%02X\",%n", t.subfunc);
                pw.printf("    \"subfunc_int\": %d,%n", t.subfunc);
                pw.printf("    \"loop_modifier\": 0,%n");
                pw.printf("    \"timer\": \"0x%08X\",%n", t.timer);
                pw.printf("    \"saved_state\": \"0x%08X\",%n", t.savedState);
                pw.printf("    \"ret_code\": \"0x%02X\",%n", t.retCode);
                pw.printf("    \"seed_hex\": \"%s\",%n", hexBytes(t.seedBytes));
                long seedBe = ((t.seedBytes[0] & 0xFFL) << 24)
                            | ((t.seedBytes[1] & 0xFFL) << 16)
                            | ((t.seedBytes[2] & 0xFFL) << 8)
                            |  (t.seedBytes[3] & 0xFFL);
                pw.printf("    \"seed_be_u32\": \"0x%08X\",%n", seedBe);
                pw.printf("    \"expected_key\": \"0x%08X\",%n", t.expectedKey);
                pw.printf("    \"state_flag\": \"0x%08X\",%n", t.stateFlag);
                pw.printf("    \"steps\": %d,%n", t.steps);
                pw.printf("    \"error\": \"%s\"%n", t.error == null ? "" : t.error);
                pw.print("  }");
                if (i < trials.size() - 1) pw.println(",");
                else pw.println();
            }
            pw.println("]");
        }
    }

    @Override
    public void run() throws Exception {
        defaultSpace = currentProgram.getAddressFactory().getDefaultAddressSpace();
        EmulatorHelper emu = new EmulatorHelper(currentProgram);
        List<Trial> trials = new ArrayList<>();
        try {
            int[] subfuncs = {0x01, 0x03, 0x05, 0x07, 0x09, 0x0B};
            long timerVal = 0;
            long POLY0 = 0xA218B900L;
            long[] seedsViaSavedState = {
                0x00000000L, 0x00000001L, 0xFFFFFFFFL, 0x80000000L,
                0x12345678L, 0xDEADBEEFL, 0xCAFEBABEL, 0xA5A5A5A5L,
                0x5A5A5A5AL, 0xAAAAAAAAL, 0x55555555L, 0x01010101L,
                0x10101010L, 0x7FFFFFFFL, 0x00000002L, 0x00000003L,
                0xA218B900L, 0x8F04FB9DL, 0xF0529CFBL, 0xAF83B3F7L,
                0x98C02CAFL, 0xE0E0B04AL, 0xFF00FF00L, 0x00FF00FFL,
                0x0F0F0F0FL, 0xF0F0F0F0L, 0xDEADC0DEL, 0xFACEFEEDL,
                0xBADC0FFEL, 0x8BADF00DL,
            };
            for (int sf : subfuncs) {
                for (long s : seedsViaSavedState) {
                    long ss = (s ^ POLY0) & 0xFFFFFFFFL;
                    Trial t = runTrial(emu, sf, timerVal, ss, 0);
                    trials.add(t);
                }
            }
        } finally {
            emu.dispose();
        }
        String outPath = "/tmp/keygen_oracle_trials.json";
        writeJson(trials, outPath);
        println(String.format("wrote %d trials -> %s", trials.size(), outPath));
        int ok = 0, faults = 0;
        for (Trial t : trials) {
            if (t.error != null && !t.error.isEmpty()) faults++;
            else ok++;
        }
        println(String.format("ran clean: %d   faulted: %d", ok, faults));
    }
}
